### PR TITLE
Introduce generic machine descriptor

### PR DESCRIPTION
**Pull request by dbaluta**
Introduce a new struct snd_soc_sof_mach that
unifies machine driver descriptor for OF/ACPI
types of devices.

Also, introduce some helper functions to set/get
the machine driver descriptor fields.

Tested on IMX8QXP works fine. Lets see if it works on Intel platforms.

### CODE REVIEW

File: sound/soc/sof/sof-of-dev.c [hunk 1]
```diff
@@ -39,6 +39,37 @@ static void sof_of_probe_complete(struct device *dev)
...
+
+	/* firmware-name is optional in DT */
+	of_property_read_string(np, "firmware-name", &sof_pdata->fw_filename);
+	ret = of_property_read_string(np, "tplg-name",
+				      &sof_pdata->tplg_filename);
+	if (ret < 0)
+		return ret;
+
+	ret = of_property_read_string(np, "machine-drv-name",
+				      &mach->drv_name);
```
Comment (plbossart) [inline sound/soc/sof/sof-of-dev.c:70]: I thought the machine driver would be probed using DT? This feels like you would still be creating a platform device using the machine driver names, which we do in ACPI since we don't have a hook to probe the machine driver, but since you do in DT why is this necessary?

Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c:70]: @plbossart yes machine driver is probed using DT. 

SOF core creates two platform devices:
  * DSP platform device
  * machine platform device.

In the OF case we only create the DSP platform device the operation creating the machine device is empty.

We still need the machine driver name here because of this:

```
snd_sof_new_platform_drv
   -> snd_sof_new_platform_drv
        pd->ignore_machine = drv_name;
```

The simplified normal code flow for DT is like this:
```
sof-of-dev.c:
-> sof_of_probe
     -> snd_sof_device_probe
         -> sof_probe_continue
            -> sof_machine_check (empty op on IMX8)
            -> snd_sof_new_platform_drv (intializes some fields of the platform driver)
            -> devm_snd_soc_register_component (registers the DSP platform driver).
            -> snd_sof_machine_register (this is an EMPTY operation on IMX8)
```

**Comment by ranj063** [comment]
@dbaluta the overall direction seems fine but there are some build errors. Could you please look into it?

**Comment by dbaluta** [comment]
> @dbaluta the overall direction seems fine but there are some build errors. Could you please look into it?

@ranj063 looking into them.

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by dbaluta** [comment]
SOFCI TEST

File: sound/soc/sof/sof-of-dev.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/sof-of-dev.c:70]: > We still need the machine driver name here because of this:
> 
> ```
> snd_sof_new_platform_drv
>    -> snd_sof_new_platform_drv
>         pd->ignore_machine = drv_name;
> ```

Why don't we remove this? I can't recall what this .ignore_machine is supposed to do.

Comment (ranj063) [reply sound/soc/sof/sof-of-dev.c:70]: @plbossart ignore_machine is very important for us. It is what tells that the FE DAI links in the machine driver should be ignored for SOF

Comment (plbossart) [reply sound/soc/sof/sof-of-dev.c:70]: @ranj063 this is only needed for legacy machine drivers with statically defined FEs, why would it matter for a new OF/DT machine driver?

Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c:70]: @ranj063 @plbossart I spent a lot of time understanding what is the purpose of ignore_machine even sent some patches to clarify the comments inside soc_check_tplg_fes.

I cannot tell you that I fully understand how to go on without this field. Besides ignoring  FE DAIs it also allows us to override BEs.

Actually, my entire machine driver descriptions starts with a regular link and soc_check_tplg_fes function converts it to a BE.

Also, this overrides link platforms name.

```  dai_link->platforms->name = component->name```

@lgirdwood maybe you can shed some light on the usage of ignore_machine.

Comment (ranj063) [reply sound/soc/sof/sof-of-dev.c:70]: @dbaluta, @plbossart is right that for new machine drv, ignore_machine is not really important (even in the case of the new ACPI machine drivers for Intel platforms ex sof_rt5682.c). What we should do is may be set this in the snd_sof_machine_select op so that we restrict it to just the Intel legacy machine drivers

File: include/sound/sof.h [hunk 2]
```diff
@@ -14,9 +14,78 @@
 #include <linux/pci.h>
 #include <sound/soc.h>
 #include <sound/soc-acpi.h>
+#include <sound/soc-of.h>
 
 struct snd_sof_dsp_ops;
 
+enum {
+	SND_SOC_SOF_MACH_TYPE_ACPI = 0,
```
Comment (lyakh) [inline include/sound/sof.h:22]: no need for initialiser

File: include/sound/sof.h [hunk 3]
```diff
@@ -14,9 +14,78 @@
...
+
+	switch (mach->type) {
+	case SND_SOC_SOF_MACH_TYPE_ACPI:
+		mach->acpi = machine;
+		break;
+	case SND_SOC_SOF_MACH_TYPE_OF:
+		mach->of = machine;
+		break;
+	default:
+		return;
```
Comment (lyakh) [inline include/sound/sof.h]: last two lines are superfluous

File: include/sound/sof.h [hunk 4]
```diff
@@ -14,9 +14,78 @@
...
 
 struct snd_sof_dsp_ops;
 
+enum {
+	SND_SOC_SOF_MACH_TYPE_ACPI = 0,
+	SND_SOC_SOF_MACH_TYPE_OF,
+};
+
+struct snd_soc_sof_mach {
+	int type;
```
Comment (lyakh) [inline include/sound/sof.h]: you could give the above enum a name and then use that type here

File: sound/soc/sof/sof-of-dev.c [hunk 5]
```diff
@@ -53,6 +53,9 @@ static int sof_of_probe(struct platform_device *pdev)
 	if (!sof_pdata)
 		return -ENOMEM;
 
+
```
Comment (lyakh) [inline sound/soc/sof/sof-of-dev.c:100]: 1 empty line is enough IMHO

File: sound/soc/sof/intel/hda.c [hunk 6]
```diff
@@ -740,15 +740,16 @@ static int hda_generic_machine_select(struct snd_sof_dev *sdev)
...
 
-			pdata->machine = hda_mach;
+			sof_mach_set_machine(pdata->machine, hda_mach,
+					     SND_SOC_SOF_MACH_TYPE_ACPI);
 			pdata->tplg_filename = tplg_filename;
 		}
 	}
 
 	/* used by hda machine driver to create dai links */
 	if (pdata->machine) {
```
Comment (lyakh) [inline sound/soc/sof/intel/hda.c:755]: this is always true now, isn't it? Is this the only check like this?

File: sound/soc/sof/sof-of-dev.c [hunk 7]
```diff
@@ -53,6 +84,10 @@ static int sof_of_probe(struct platform_device *pdev)
 	if (!sof_pdata)
 		return -ENOMEM;
 
+	sof_pdata->machine = devm_kzalloc(dev, sizeof(*sof_pdata->machine),
+					  GFP_KERNEL);
+	if (!sof_pdata->machine)
+		return -ENOMEM;
```
Comment (lyakh) [inline sound/soc/sof/sof-of-dev.c:99]: ok, you re-purpose your double empty line here :-)

File: include/sound/sof.h (hunk 4 above)
Comment (dbaluta) [reply include/sound/sof.h]: good point. will do it with next version.

File: sound/soc/sof/sof-of-dev.c (hunk 7 above)
Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c:99]: @lyakh I understand now! Thanks for noticing this :). Will fix it in next version.

File: sound/soc/sof/sof-of-dev.c [hunk 8]
```diff
@@ -75,6 +110,12 @@ static int sof_of_probe(struct platform_device *pdev)
 	sof_pdata->fw_filename_prefix = sof_pdata->desc->default_fw_path;
 	sof_pdata->tplg_filename_prefix = sof_pdata->desc->default_tplg_path;
 
+	ret = sof_of_parse(pdev);
+	if (ret < 0) {
+		dev_err(dev, "error: failed to parse DSP DT node\n");
```
Comment (RanderWang) [inline sound/soc/sof/sof-of-dev.c:124]: can you also print ret in debug message ?

File: sound/soc/sof/sof-of-dev.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/sof-of-dev.c:70]: @dbaluta if we avoid this ignore_machine thingy then we don't need all these definitions... I am lazy but less code is better in most cases :-)

Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c:70]: @plbossart agree with you on that. Will try to remove it and adapt simple-audio-card for this new situation.

Using ignore_machine I was able to create one normal link with simple-audio card and convert it to a BE inside soc_check_tplg_fes :). And everything worked really fine.

File: sound/soc/sof/sof-of-dev.c (hunk 8 above)
Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c:124]: @RanderWang good point.

File: include/sound/sof.h (hunk 2 above)
Comment (dbaluta) [reply include/sound/sof.h:22]: I've seen this initializer in a lot of places. It gives the reader a sense of what's there. But yes will remove it. 

After @plbossart comments I think removing ignore_machine name from the common path will drastically change this PR :).

**Comment by plbossart** [comment]
@dbaluta do you plan on updating this PR or rather to fix the ignore_machine thingy?
I've done other SoundWire things and I don't see a clear direction.

**Comment by dbaluta** [comment]
@plbossart  if Intel doesn't plan to deprecate ignore_machine thingy I would rather update this PR. 
simple-card machine driver is in the same situation as the older Intel drivers. It can only create normal links OR FE/BE links where for FE links codec is *always* dummy and for BE links cpu is *always* dummy. 

In order to eliminate ignore_machine thingy simple-card cannot really be used, so we need to write a new machine driver. I would prefer not to do that right.

So, my proposal would be to update this PR and keep the ignore_machine even for ARM. Will update this today. Sorry for being late I had some priority tasks and also spent some time on investigating how to enhance simple-card, but its basically rewriting it. Would also ask Morimoto-san about this.

File: sound/soc/sof/sof-of-dev.c (hunk 8 above)
Comment (dbaluta) [reply sound/soc/sof/sof-of-dev.c:124]: @RanderWang actually i don't think is needed. Probe when fails it will return the error stored in ret.

**Comment by dbaluta** [comment]
@plbossart i have updated the PR.

**Comment by plbossart** [comment]
> @plbossart if Intel doesn't plan to deprecate ignore_machine thingy I would rather update this PR.
> simple-card machine driver is in the same situation as the older Intel drivers. It can only create normal links OR FE/BE links where for FE links codec is _always_ dummy and for BE links cpu is _always_ dummy.
> 
> In order to eliminate ignore_machine thingy simple-card cannot really be used, so we need to write a new machine driver. I would prefer not to do that right.
> 
> So, my proposal would be to update this PR and keep the ignore_machine even for ARM. Will update this today. Sorry for being late I had some priority tasks and also spent some time on investigating how to enhance simple-card, but its basically rewriting it. Would also ask Morimoto-san about this.

I am not happy about maintaining something none of us quite understands because of legacy issues and the simpe-card design. Let's try and remove the things we don't need.

**Comment by dbaluta** [comment]
Ok.

**Comment by dbaluta** [comment]
Closing this because we first need to move out ignore_machine from common path first.
